### PR TITLE
Fixes various typing errors in `pytorch_lightning/strategies/deepspeed.py`

### PR DESCRIPTION
**Pull request by donlapark**
## What does this PR do?

Fixes mypy errors in `pytorch_lightning/strategies/deepspeed.py` as addressed in #13445.

The code suggests that `lightning_model.training` is always assumed to have type `pl.Trainer` and not `None`. However, `mypy` does not know this and throw out an error whenever `lightning_model.training` is called.

In addition, `mypy` seems to not recognize `hasattr` as mentioned in [this issue](https://github.com/python/mypy/issues/1424), causing it to throw an error on the following line (671-672) without `#type: ignore`:

```
if hasattr(train_dataloader, "batch_sampler"):
    batch_size = train_dataloader.batch_sampler.batch_size  # type: ignore[union-attr]
```

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Comment by otaj** [comment]
Hi @donlapark, it seems, that if you remove `"pytorch_lightning.strategies.deepspeed"` from `pyproject.toml` file, not many errors go away. I pushed the change with removal of the file from mypy ignoring so you can easily see them.

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 1]
```diff
@@ -396,7 +403,9 @@ def _set_node_environment_variables(self) -> None:
     def restore_checkpoint_after_setup(self) -> bool:
         return True
 
-    def _setup_model_and_optimizers(self, model: Module, optimizers: List[Optimizer]) -> Tuple[Module, List[Optimizer]]:
+    def _setup_model_and_optimizers(
+        self, model: Module, optimizers: List[Optimizer]
+    ) -> Tuple[DeepSpeedEngine, List[Optimizer]]:
```
Comment (otaj) [inline src/pytorch_lightning/strategies/deepspeed.py:408]: Hi @donlapark, this seems to break tests at runtime as `DeepSpeedEngine` is not available. https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING

Comment (donlapark) [reply src/pytorch_lightning/strategies/deepspeed.py]: Thanks! Let's see if `"deepspeed.DeepSpeedEngine"` works.

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 2]
```diff
@@ -331,7 +331,7 @@ def __init__(
         self.hysteresis = hysteresis
         self.min_loss_scale = min_loss_scale
 
-    def _load_config(self, config):
+    def _load_config(self, config: Optional[Union[Path, str, Dict[str, Any]]]) -> Optional[Dict[str, Any]]:
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/deepspeed.py:334]: can you `_PATH` from `utilities.types`.

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 3]
```diff
@@ -359,6 +360,7 @@ def setup_distributed(self):
     def setup(self, trainer: "pl.Trainer") -> None:
         self.accelerator.setup(trainer)
         # we set the device so that optimizers can be created with distributed comms.
+        assert isinstance(self.lightning_module, pl.LightningModule)
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/deepspeed.py:363]:
```suggestion
        assert self.lightning_module is not None
```
I think it should check the other case.

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 4]
```diff
@@ -367,6 +369,7 @@ def setup(self, trainer: "pl.Trainer") -> None:
         self.barrier()
 
     def _init_deepspeed_distributed(self) -> None:
+        assert isinstance(self.cluster_environment, ClusterEnvironment)
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/deepspeed.py:372]: same here.

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 5]
```diff
@@ -378,14 +381,15 @@ def _init_deepspeed_distributed(self) -> None:
...
-    def _get_process_group_backend(self):
+    def _get_process_group_backend(self) -> str:
         return (
             self._process_group_backend
             or _get_process_group_backend_from_env()
             or get_default_process_group_backend_for_device(self.root_device)
         )
 
     def _set_node_environment_variables(self) -> None:
+        assert isinstance(self.cluster_environment, ClusterEnvironment)
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/deepspeed.py:392]: same here

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 6]
```diff
@@ -405,6 +411,7 @@ def _setup_model_and_optimizers(self, model: Module, optimizers: List[Optimizer]
             The model wrapped into a :class:`deepspeed.DeepSpeedEngine` and a list with a single
             deepspeed optimizer.
         """
+        assert isinstance(self.config, dict)
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/deepspeed.py:414]: same here. also I'd recommend adding assertions right before it is used.

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 7]
```diff
@@ -642,16 +670,18 @@ def _format_batch_size_and_grad_accum_config(self):
         if "gradient_clipping" not in self.config:
             self.config["gradient_clipping"] = self.lightning_module.trainer.gradient_clip_val or 0.0
 
-    def _auto_select_batch_size(self):
+    def _auto_select_batch_size(self) -> Optional[int]:
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/deepspeed.py:673]: when is it optional?

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 8]
```diff
@@ -914,14 +952,17 @@ def register_strategies(cls, strategy_registry: Dict) -> None:
...
 
-    def test_step(self, *args, **kwargs) -> Optional[STEP_OUTPUT]:
+    def test_step(self, *args: Any, **kwargs: Any) -> Optional[STEP_OUTPUT]:
+        assert self.model is not None
         with self.precision_plugin.test_step_context():
             return self.model(*args, **kwargs)
 
-    def predict_step(self, *args, **kwargs) -> STEP_OUTPUT:
+    def predict_step(self, *args: Any, **kwargs: Any) -> STEP_OUTPUT:
+        assert self.model is not None
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/deepspeed.py:966]: i'd suggest waiting for #13519

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 9]
```diff
@@ -48,12 +48,12 @@
...
 from pytorch_lightning.utilities.rank_zero import rank_zero_info
 from pytorch_lightning.utilities.seed import reset_seed
-from pytorch_lightning.utilities.types import _PATH, LRSchedulerConfig, LRSchedulerTypeUnion, STEP_OUTPUT
+from pytorch_lightning.utilities.types import _LRScheduler, _PATH, LRSchedulerConfig, ReduceLROnPlateau, STEP_OUTPUT
 from pytorch_lightning.utilities.warnings import rank_zero_warn, WarningCache
 
 warning_cache = WarningCache()
 
-_DEEPSPEED_AVAILABLE: bool = _RequirementAvailable("deepspeed")
+_DEEPSPEED_AVAILABLE: bool = cast(bool, _RequirementAvailable("deepspeed"))
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/deepspeed.py:56]: This should work

```suggestion
_DEEPSPEED_AVAILABLE = _RequirementAvailable("deepspeed")
```

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 10]
```diff
@@ -431,14 +441,16 @@ def _setup_model_and_optimizer(
...
             optimizer=optimizer,
             lr_scheduler=lr_scheduler,
             dist_init_required=False,
         )
         return deepspeed_engine, deepspeed_optimizer
 
-    def init_deepspeed(self):
+    def init_deepspeed(self) -> None:
+        assert isinstance(self.lightning_module, pl.LightningModule)
+        assert isinstance(self.lightning_module.trainer, pl.Trainer)
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/deepspeed.py:453]: This is not necessary anymore after #13638 

```suggestion
```

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 11]
```diff
@@ -799,10 +830,13 @@ def load_checkpoint(self, checkpoint_path: _PATH) -> Dict[str, Any]:
         if self.load_full_weights and self.zero_stage_3:
             # Broadcast to ensure we load from the rank 0 checkpoint
             # This doesn't have to be the case when using deepspeed sharded checkpointing
-            checkpoint_path = self.broadcast(checkpoint_path)
+            checkpoint_path = cast(_PATH, self.broadcast(checkpoint_path))
             return super().load_checkpoint(checkpoint_path)
 
         # Rely on deepspeed to load the checkpoint and necessary information
+        assert isinstance(self.lightning_module, pl.LightningModule)
+        assert isinstance(self.lightning_module.trainer, pl.Trainer)
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/deepspeed.py:838]:
```suggestion
```

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 12]
```diff
@@ -818,6 +852,8 @@ def load_checkpoint(self, checkpoint_path: _PATH) -> Dict[str, Any]:
 
     @property
     def lightning_restore_optimizer(self) -> bool:
+        assert isinstance(self.lightning_module, pl.LightningModule)
+        assert isinstance(self.lightning_module.trainer, pl.Trainer)
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/deepspeed.py:856]:
```suggestion
```

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 13]
```diff
@@ -642,16 +670,18 @@ def _format_batch_size_and_grad_accum_config(self):
         if "gradient_clipping" not in self.config:
             self.config["gradient_clipping"] = self.lightning_module.trainer.gradient_clip_val or 0.0
 
-    def _auto_select_batch_size(self):
+    def _auto_select_batch_size(self) -> Optional[int]:
         # train_micro_batch_size_per_gpu is used for throughput logging purposes
         # by default we try to use the batch size of the loader
+        assert isinstance(self.lightning_module, pl.LightningModule)
+        assert isinstance(self.lightning_module.trainer, pl.Trainer)
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/deepspeed.py:677]:
```suggestion
```

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 14]
```diff
@@ -616,20 +642,22 @@ def handles_gradient_accumulation(self) -> bool:
...
         self._format_precision_config()
 
-    def _format_batch_size_and_grad_accum_config(self):
+    def _format_batch_size_and_grad_accum_config(self) -> None:
         # todo: using lite, we do not support these variables within the config
+        assert isinstance(self.config, dict)
         if self.lightning_module is None:
             return
 
+        assert isinstance(self.lightning_module.trainer, pl.Trainer)
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/deepspeed.py:660]:
```suggestion
```

File: src/pytorch_lightning/strategies/deepspeed.py (hunk 1 above)
Comment (donlapark) [reply src/pytorch_lightning/strategies/deepspeed.py]: Alright, it works!

File: src/pytorch_lightning/strategies/deepspeed.py (hunk 2 above)
Comment (donlapark) [reply src/pytorch_lightning/strategies/deepspeed.py]: Done!

File: src/pytorch_lightning/strategies/deepspeed.py (hunk 9 above)
Comment (donlapark) [reply src/pytorch_lightning/strategies/deepspeed.py]: It does not work :(. `mypy` throws `error: Cannot determine type of "_DEEPSPEED_AVAILABLE"  [has-type]`.

Comment (donlapark) [reply src/pytorch_lightning/strategies/deepspeed.py]: So the error is actually for a different file (`src/pytorch_lightning/plugins/precision/deepspeed.py`). I have applied a quick fix to the said file and this should resolve the problem.

File: src/pytorch_lightning/strategies/deepspeed.py (hunk 7 above)
Comment (donlapark) [reply src/pytorch_lightning/strategies/deepspeed.py]: Yeah, this should not be optional. Fixed and thanks!

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 15]
```diff
@@ -331,7 +331,7 @@ def __init__(
         self.hysteresis = hysteresis
         self.min_loss_scale = min_loss_scale
 
-    def _load_config(self, config):
+    def _load_config(self, config: Optional[Union[_PATH, str, Dict[str, Any]]]) -> Optional[Dict[str, Any]]:
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/deepspeed.py:334]:
```suggestion
    def _load_config(self, config: Optional[Union[_PATH, Dict[str, Any]]]) -> Optional[Dict[str, Any]]:
```

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 16]
```diff
@@ -431,14 +442,15 @@ def _setup_model_and_optimizer(
...
+            model_parameters=model_parameters,
             optimizer=optimizer,
             lr_scheduler=lr_scheduler,
             dist_init_required=False,
         )
         return deepspeed_engine, deepspeed_optimizer
 
-    def init_deepspeed(self):
+    def init_deepspeed(self) -> None:
+        assert isinstance(self.lightning_module, pl.LightningModule)
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/deepspeed.py:453]:
```suggestion
        assert self.lightning_module is not None
```

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 17]
```diff
@@ -464,6 +476,7 @@ def init_deepspeed(self):
                 "DeepSpeed currently does not support different `accumulate_grad_batches` at different epochs."
             )
 
+        assert isinstance(self.model, (pl.LightningModule, _LightningPrecisionModuleWrapperBase))
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/deepspeed.py:479]:
```suggestion
        assert self.model is not None
```

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 18]
```diff
@@ -472,6 +485,7 @@ def init_deepspeed(self):
             self._initialize_deepspeed_inference(model)
 
     def _init_optimizers(self) -> Tuple[Optimizer, Optional[LRSchedulerConfig], Optional[int]]:
+        assert isinstance(self.lightning_module, pl.LightningModule)
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/deepspeed.py:488]:
```suggestion
        assert self.lightning_module is not None
```

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 19]
```diff
@@ -642,16 +669,17 @@ def _format_batch_size_and_grad_accum_config(self):
         if "gradient_clipping" not in self.config:
             self.config["gradient_clipping"] = self.lightning_module.trainer.gradient_clip_val or 0.0
 
-    def _auto_select_batch_size(self):
+    def _auto_select_batch_size(self) -> int:
         # train_micro_batch_size_per_gpu is used for throughput logging purposes
         # by default we try to use the batch size of the loader
+        assert isinstance(self.lightning_module, pl.LightningModule)
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/deepspeed.py:675]:
```suggestion
        assert self.lightning_module is not None
```

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 20]
```diff
@@ -799,10 +828,12 @@ def load_checkpoint(self, checkpoint_path: _PATH) -> Dict[str, Any]:
         if self.load_full_weights and self.zero_stage_3:
             # Broadcast to ensure we load from the rank 0 checkpoint
             # This doesn't have to be the case when using deepspeed sharded checkpointing
-            checkpoint_path = self.broadcast(checkpoint_path)
+            checkpoint_path = cast(_PATH, self.broadcast(checkpoint_path))
             return super().load_checkpoint(checkpoint_path)
 
         # Rely on deepspeed to load the checkpoint and necessary information
+        assert isinstance(self.lightning_module, pl.LightningModule)
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/deepspeed.py:835]:
```suggestion
        assert self.lightning_module is not None
```

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 21]
```diff
@@ -818,6 +849,7 @@ def load_checkpoint(self, checkpoint_path: _PATH) -> Dict[str, Any]:
 
     @property
     def lightning_restore_optimizer(self) -> bool:
+        assert isinstance(self.lightning_module, pl.LightningModule)
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/deepspeed.py:852]:
```suggestion
        assert self.lightning_module is not None
```

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 22]
```diff
@@ -842,11 +874,13 @@ def _restore_zero_state(self, ckpt: Mapping[str, Any]) -> None:
             ckpt: The ckpt file.
         """
 
-        def load(module: torch.nn.Module, prefix=""):
+        assert isinstance(self.lightning_module, pl.LightningModule)
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/deepspeed.py:877]:
```suggestion
        assert self.lightning_module is not None
```

File: src/pytorch_lightning/strategies/deepspeed.py (hunk 15 above)
Comment (rohitgr7) [reply src/pytorch_lightning/strategies/deepspeed.py:334]: also, mind updating the typing for config in init?

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 23]
```diff
@@ -485,10 +499,18 @@ def _init_optimizers(self) -> Tuple[Optimizer, Optional[LRSchedulerConfig], Opti
...
     def zero_stage_3(self) -> bool:
-        return self.config.get("zero_optimization") and self.config.get("zero_optimization").get("stage") == 3
+        assert isinstance(self.config, dict)
+        zero_optimization = self.config.get("zero_optimization")
+        if zero_optimization is None:
+            raise MisconfigurationException(
+                "You must set `zero_optimization` in the DeepSpeed config."
+                " See: https://pytorch-lightning.readthedocs.io/en/stable/advanced/model_parallel.html#deepspeed"
+            )
+        return bool(self.config.get("zero_optimization")) and zero_optimization.get("stage") == 3
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/deepspeed.py:509]: should we raise Exceptions inside properties? Ideally, it should just check and return False if not found.
```suggestion
        assert isinstance(self.config, dict)
        zero_optimization = self.config.get("zero_optimization")
        return zero_optimization and zero_optimization.get("stage") == 3
```

Comment (donlapark) [reply src/pytorch_lightning/strategies/deepspeed.py:509]: We'd probably not. Thanks!

File: src/pytorch_lightning/strategies/deepspeed.py (hunk 15 above)
Comment (donlapark) [reply src/pytorch_lightning/strategies/deepspeed.py]: Sure. Done!

File: src/pytorch_lightning/strategies/deepspeed.py [hunk 24]
```diff
@@ -485,10 +499,13 @@ def _init_optimizers(self) -> Tuple[Optimizer, Optional[LRSchedulerConfig], Opti
 
     @property
     def zero_stage_3(self) -> bool:
-        return self.config.get("zero_optimization") and self.config.get("zero_optimization").get("stage") == 3
+        assert isinstance(self.config, dict)
+        zero_optimization = self.config.get("zero_optimization")
+        return zero_optimization and zero_optimization.get("stage") == 3
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/deepspeed.py:504]:
```suggestion
        return zero_optimization is not None and zero_optimization.get("stage") == 3
```

**Review by rohitgr7 — APPROVED** [review]
Great work!

BTW, you can also add all suggestions in a batch and commit them at once inside the PR's `Files changed` section.

ref: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request

**Comment by donlapark** [comment]
Wow, I should have known that. Thanks for the tip!

**Review by otaj — APPROVED** [review]

File: src/pytorch_lightning/strategies/deepspeed.py (hunk 8 above)
Comment (awaelchli) [reply src/pytorch_lightning/strategies/deepspeed.py:957]: #13519 was merged, we can update this now.

**Review by awaelchli — APPROVED** [review]
Great!

**Pull request merged**